accepted/0000-selective-versions-resolutions.md

```diff
@@ -0,0 +1,495 @@
+    "typescript": "2.3.2"
+  },
+  "resolutions": {
+    "typescript": "2.0.2"
+  }
+```
+
+yarn would use `typescript@2.0.2` for the whole project and that's all.
+The same kind of consideration (outside of the glob pattern thing) should be
+followed as with the selected solution of this RFC.
+
+This is basically too simple according to discussions with yarn maintainers.
+
+## Mapping version specifications
+
+This is a kind of simplified solution to the "out-of-scope scenario" presented
+in the Motivations section above (it maps versions but not dependency names).
+
+It was proposed in this
+[comment](https://github.com/yarnpkg/yarn/issues/2763#issuecomment-301896274).
+
+It is similar to the previous alternative but with a version specification
+in the package designation. This would take this form in the `package.json`:
+```json
+  "devDependencies": {
+    "@angular/cli": "1.0.3",
+    "typescript": "2.2.2",
+    "more dependencies..."
+  },
+  "resolutions": {
+    "typescript@>=2.0.0 <2.3.0": "typescript@2.3.2"
+  }
+```
+
+yarn would then replace matching version specifications with the user's one.
+For example a dependency normally resolved to `typescript@2.2.2` would be
+resolved in practice to `typescript@2.3.2`.
+
+This is too advanced and can be considered a possible extension of this RFC.
+
+## Mapping version specifications as well as packages name
+
+Same as the two above but with a different name on the right-hand side of the
+*resolution*:
+```json
+  "devDependencies": {
+    "@angular/cli": "1.0.3",
+    "typescript": "2.2.2"
+  },
+  "resolutions": {
+    "typescript@>=2.0.0 <2.3.0": "my-typescript-fork@2.3.2"
+  }
+```
+
+or even:
+```json
+  "devDependencies": {
+    "@angular/cli": "1.0.3",
+    "typescript": "2.2.2"
+  },
+  "resolutions": {
+    "typescript": "my-typescript-fork",
+  }
+```
+
+and the version specification would be conserved.
+
+This is too advanced and can be considered a possible extension of this RFC.
+
+# Future extensions
+
+The two alternatives discussed in the section just above, "Mapping version
+specifications" and "Mapping version specifications as well as packages name",
+can be adapted to the current proposition to support these uses cases as well.
+
+# Unresolved questions
+
+## Package designation and single `*`
+
+I am not totally convinced that using a single `*` in a package designation
+is a good idea. Mostly because it introduces much for uncertainty to what it is
+going to be applied. In particular, I'm thinking about defining a *resolution*
+like that and latter adding a dependency which has a dependency that matches
+but is not intended to. I feel like it adds a lot of mental weight for the
+user to manipulate single `*`.
+
+## `--flat` option
+
+I wonder if the `--flat` option of `install` shouldn't be transformed to
+a `flatten` command that would:
+1. Fill in the *resolutions* for all nested dependencies.
+2. Set the `flat` field in the `package.json`.

```

Comment:
I like that idea.
Unlike --production, --flat is an option that you decide for the project early on and stick to it.